component/src/lib.rs

```diff
@@ -0,0 +1,127 @@
+///
+/// 2. A handle_irrecoverable which takes actions on a relaunch due to an irrecoverable error
+/// that happened. It takes the error message that may contain a stack trace and other information
+/// that was sent to the Supervisor via the tx_irrecoverable passed into start.
+#[async_trait]
+pub trait Manageable {
+    // The function that spawns a tokio task
+    async fn start(
+        &self,
+        tx_irrecoverable: Sender<anyhow::Error>,
+        rx_cancellation: oneshotReceiver<()>,
+    ) -> tokio::task::JoinHandle<()>; // Note the task is "silent" (returns nothing)
+
+    // The function for cleanup after the task has encountered an irrecoverable error
+    fn handle_irrecoverable(&self, irrecoverable: IrrecoverableError) -> Result<(), anyhow::Error>;
+}
+
+impl<M: Manageable + Send> Supervisor<M> {
+    /// Creates a new supervisor using a Manageable component.
+    pub fn new(component: M) -> Self {
+        let (_, tr_irrecoverable) = channel(10);
+        let (tx_cancellation, _) = oneshotChannel();
+        Supervisor {
+            irrecoverable_signal: tr_irrecoverable,
+            cancellation_signal: Some(tx_cancellation),
+            join_handle: None,
+            manageable: component,
+        }
+    }
+
+    /// Spawn calls the start function of the Manageable component and runs supervision.
+    pub async fn spawn(mut self) -> Result<(), anyhow::Error> {
+        let (tx_irrecoverable, tr_irrecoverable) = channel(10);
+        let (tx_cancellation, tr_cancellation) = oneshotChannel();
+
+        // this assumes the start_fn is populated
+        let wrapped_handle = self
+            .manageable
+            .start(tx_irrecoverable, tr_cancellation)
+            .await;
+
+        self.irrecoverable_signal = tr_irrecoverable;
+        self.cancellation_signal = Some(tx_cancellation);
+        self.join_handle = Some(wrapped_handle);
+
+        self.run().await
+    }
+
+    /// Run watches continuously for irrecoverable errors or JoinHandle completion.
+    pub async fn run(mut self) -> Result<(), anyhow::Error> {
+        // select statement that listens for the following cases:
+        //
+        // Irrecoverable signal incoming => log, terminate and restart
+        // completion of the task => already terminated, log and restart
+        //
+        // The handle_irrecoverable is run before the existing task gets
+        // cancelled by restart in the case that an irrecoverable signal
+        // was sent to us. This makes resource cleanup possible.
+        loop {
+            tokio::select! {
+                Some(message) = self.irrecoverable_signal.recv() => {
+                    self.manageable.handle_irrecoverable(message)?;
+                    self.restart().await?;
+                },
+
+               // Poll the JoinHandle<O>
+               _result =  self.join_handle.as_mut().unwrap(), if self.join_handle.is_some() => {
```

Comment:
Why is this unwrap safe?

Reply:
How about some pattern matching like this:
Ok(result) = self.join_handle.as_mut()